blogposts/2020/code-search-turned-code-checker.md

```diff
@@ -0,0 +1,258 @@
+about. I relied only on patterns in `staticcheck`'s test data to discover syntactic
+variants for checks, so I don't know if I covered all the patterns that
+`staticcheck` implements in its code.
+
+The majority of checks that I couldn't write required type information to
+implement correctly (13 of 34). Other checks I couldn't write required more
+complex syntax matching rules (8 of 34). This table roughly quantifies the
+expressive needs for implementing checks:
+
+| Description                      | # of checks |
+|----------------------------------|-------------|
+| Total                            | 34          |
+| Works (all variants)             | 11          |
+| Works (at least one variant)     | 19          |
+| Needs additional syntax matching | 8           |
+| Needs type info                  | 13          |
+
+<sup>Note that some checks require type info _and_ additional syntax matching.
+Also, one working variant for a check may not require type info, but another
+variant for the same check may. I.e., the values overlap and do not sum to
+Total.</sup>
+
+Extending search queries to access static properties like type information is a
+natural extension for writing better code checks, and an area of [code intelligence](https://docs.sourcegraph.com/user/code_intelligence)
+work that we're exploring.
+
+### Takeaway
+
+Code search can be much more than finding your favorite provider called
+`BananaNut.*Cake`. It can be a lightweight workflow for revealing
+short-and-sweet ways to make your code better. Perhaps the most powerful idea is
+that a universal code search can wholesale find and eradicate the code slips
+we're always bound to make.
+
+Starting small, the solution could look like a search query that finds issues in
+active and popular Go projects at just the push of a button. Why not start
+there?
+
+[🔘 Run the entire code check query on popular Go code projects ↗](TODO)
+
+### Proposing changes
+
+The most exciting part about finding real hits with any code checking tool is
+that we discover an _actionable_ way to improve the code! I've to narrowed down
+the search query above so that it (hopefully) finds only real and
+uncontroversial patterns to improve and contribute to active Go projects.
+
+Part of an effective contribution means that we need to avoid matches in files
+that are tests, vendored, or external dependencies. While anyone can use the
+results to make contributions, it's important to be mindful of contributor
+guidelines, clearly communicate and motivate proposed changes in pull requests,
+validate that the change passes a project's tests or CI checks. And, while the
+query does exclude common test and vendored files, it's best to check that
+matches occur in files that really are part of the project.
+
+If you're interested in potentially turning these results into open source
+contributions, e-mail me at <rijnard@sourcegraph.com>, I can help.
```

Comment:
This is an idea and still open-ended :-) Will tag and run it by Beyang and others when the post is in a more complete draft mode.